### ISSUE TITLE
[1.4.3] Anoymous default exports not possible

### ISSUE THREAD

**Issue by jbbr** [meteor/meteor#8312]
I wanted to test latest 1.4.2.5-rc.0 and 1.4.3-rc.0 (`ecmascript` 0.6.2-rc.0) in a project of mine and got babel errors:
```
/path/to/project/.meteor/packages/ecmascript/.0.6.2-rc.0.u2123h++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/babel-types/lib/definitions/index.js:109:13:
   /server/myclass.js: Property right of AssignmentExpression expected
   node to be of a type ["Expression"] but instead got
   "VariableDeclaration"
   at Object.validate
```

`server/myclass.js` is very basic:
```
export default class { }
```
Workaround: Giving the class a name.

Exporting anonymous functions does still work fine.

This bug also extends to Meteor packages. e.g https://github.com/cult-of-coders/grapher does not compile due to this.

If useful I can provide an example repo for reproducing. But steps are very easy:

1. `$ meteor create --release 1.4.3-rc.0 testproject`
2. Replace server/main.js with `export default class { }`
3. `$ meteor`

Update: The problem seems to be in 1.4.3+ - I appeared in 1.4.2.x too but probably after downgrading from 1.4.3 or after using the (not recommended) 1.4.2.4

**Comment by jbbr** [comment]
In `1.4.2.5` code and packages which used to work just fine `1.4.2.3` are still broken due to this bug

**Comment by klaussner** [comment]
This could have something to do with the fact that `babel-compiler` [runs Reify](https://github.com/meteor/babel/blob/master/index.js#L52-L59) (for `import` and `export`) before Babel. For your example, Reify outputs an AST for `module.export("default",exports.default=(class {}));` but Babel fails to transform it. Interestingly, passing that same (reified) code to `babel-compiler` works, so I compared both ASTs: they are identical with the exception that Reify generates a `ClassDeclaration` on the right hand side of the assignment and Babel generates a `ClassExpression`.

![bildschirmfoto 2017-02-06 um 23 59 02](https://cloud.githubusercontent.com/assets/1312807/22670385/27623d68-ecc9-11e6-9707-d108e378b183.png)

I'm not sure if or how this is related to the `VariableDeclaration` mentioned in the error message but maybe it helps anyway. :)

**Comment by abernix** [comment]
Thanks for taking a look at this, @klaussner.  I haven't looked at more than the above information, but `ClassDeclaration` _would_ seem wrong on the right side of an assignment – I think it should be a `ClassExpression`.

It's probably worth pointing out that some major things changed between the two versions listed in your subject and they use different versions of `ecmascript` which leads down the following paths:

* Meteor 1.4.2.5
  * `ecmascript@0.6.1` => `babel-compiler@6.13.0` =>
 `meteor-babel@0.13.0` (npm)
* Meteor 1.4.3
  * `ecmascript@0.6.2` => `babel-compiler@6.14.x` =>
`meteor-babel@0.14.3` (npm)

I point this out because [this commit](https://github.com/meteor/babel/commit/c8815d26f184d2dcae7bdfbeb2190b55337e9092) changed between 0.13 and 0.14.3 (and Meteor 1.4.2.x and 1.4.3, respectively).

@klaussner Given the above differences, which environment did you triage under?  Are you saying that [pre-Babel](https://github.com/meteor/babel/blob/4801669a141bc292a09bf0ab496fdfae9aaef91e/index.js#L37-L44), `reify` is returning the AST (i.e. `reifyResult.ast`) shown on the left of the diff image above?

**Comment by klaussner** [comment]
Yes, the ASTs are both from `reifyResult` directly after `reifyCompiler.compile` in 1.4.3 but for different inputs. The left one is for `export default class {}` and the right one is for `module.export("default",exports.default=(class {}));`.

**Comment by jbbr** [comment]
@abernix thanks for the overview of the dependency tree. 

I used to have this problem in some 1.4.2.x release. Probably it was only the case because I might have downgraded the release from 1.4.3.x or used the revoked 1.4.2.4.
Not sure but at least now I successfully upgraded a project to 1.4.2.6.

Latest 1.4.3 (no rc label anymore) still has this problem for me and breaks my project (because of Meteor packages using this language feature).

Is there anything I can do to help fixing this? Unfortunately I have no experience with babel's internals.

**Comment by abernix** [comment]
You might have seen outside 1.4.3 if you updated to a `1.4.3` beta/RC and then downgraded again to pre-1.4.3 without ensuring that your `babel-compiler` or `ecmascript` package(s) was/were also downgraded (version control can help avoid this).

That said, I think this could be a legit problem in 1.4.3 due to `reify` now running before `babel` and `babel` seems more stringent than say, `acorn`, when it comes to using this modified AST.  It would seem that `babel` is simply less willing to deal with a `ClassDeclaration` on the right side of an assignment than it is with a `FunctionDeclaration`.  A [test for this](https://github.com/benjamn/reify/blob/master/test/tests.js#L172-L178) exists but must be skipped under Node 4 as it doesn't support block-level declarations like `class` natively outside strict-mode.

I think `reify` can account for this, and I've opened a PR with `reify`.  While my `reify` PR does fix this problem, it probably warrants some further discussion as I'm not an expert in this area.

**Comment by abernix** [comment]
Re-opening until this surfaces via `meteor-babel` getting published ("slash" maybe just Meteor 1.4.3.1).

**Comment by abernix** [comment]
@jbbr Thank you very much for reporting this.  This should be fixed if you update to Meteor 1.4.2.7:

    meteor update --release 1.4.2.7

@klaussner Thank you for your very helpful debugging! 🥂
